src/librustc/sir.rs

```diff
@@ -0,0 +1,136 @@
+//! Serialised Intermediate Representation (SIR).
+//!
+//! SIR is built in-memory during LLVM code-generation, and finally placed into a dedicated ELF
+//! section at link time.
+
+use std::default::Default;
+use std::io::{self, Write};
+use rustc_index::{newtype_index, vec::{Idx, IndexVec}};
+use rustc_data_structures::fx::FxHashMap;
+use ykpack;
+
+// Duplicates of LLVM types defined elsewhere, copied to avoid cyclic dependencies. Whereas the
+// LLVM backend expresses pointers to these using references, we use raw pointers so to as avoid
+// introducing lifetime parameters to the SirCx (and thus into TyCtxt and every place that uses
+// it).
+extern { pub type Value; }
+extern { pub type BasicBlock; }
+
+newtype_index! {
+    pub struct SirFuncIdx {
+        DEBUG_FORMAT = "SirFuncIdx({})"
+    }
+}
+
+// The index of a block within a function.
+// Note that these indices are not globally unique. For a globally unique block identifier, a
+// (SirFuncIdx, SirBlockIdx) pair must be used.
+newtype_index! {
+    pub struct SirBlockIdx {
+        DEBUG_FORMAT = "SirBlockIdx({})"
+    }
+}
+
+/// Sir equivalents of LLVM values.
+#[derive(Debug)]
+pub enum SirValue {
+    Func(SirFuncIdx),
+}
+
+impl SirValue {
+    pub fn func_idx(&self) -> SirFuncIdx {
+        let Self::Func(idx) = self;
+        *idx
+    }
+}
+
+pub struct SirCx {
+    /// Maps an opaque LLVM `Value` to its SIR equivalent.
+    pub llvm_values: FxHashMap<*const Value, SirValue>,
+    /// Maps an opaque LLVM `BasicBlock` to the function and block index of it's SIR equivalent.
+    pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
+    /// Function store. Also owns the blocks
+    pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+}
+
+impl SirCx {
```

Comment:
Do any of the functions in this `impl` have any non-obvious constraints that need documenting? `add_func` sounds pleasingly obvious but maybe, I dunno, "the user must ensure no otehr function with this name exists" or whatever?

Reply:
Ah yes. There is a constraint, but I'll have to look up the rules. We inherit some oddity from LLVM as I recall.

Reply:
Here's what I was thinking of:
https://github.com/vext01/ykrustc/blob/083637e167e83643836750a68fb9691d189aaa82/src/librustc_codegen_llvm/declare.rs#L27-L30

That behaviour is inherited from LLVM. The upshot for us is that if the codegen defines the same symbol twice (in the same codegen) then `add_func()` will recieve the same LLVM pointer twice.

I can tell you that this doesn't happen currently, as we have an assertion that would crash out if we tried to insert into the same slot in our hash table:
https://github.com/vext01/ykrustc/blob/c42235b1b39d515ec4cb619a8923504c8d6cd63d/src/librustc/sir.rs#L75-L76

What would you like to do? We can either leave the assertion for educational purposes, or we can allow the insertion to overwrite the slot.

Reply:
I have no opinions on what we should do: I just want to make sure that a) we've thought about it b) where documentation / asserts are needed, we have added them. I'm still unsure about b) in this instance.

Reply:
What we currently have is conservative, in that, in doubt, it crashes. I think that's a good starting point.

Reply:
OK. So then my question is: do we need to document this, or is the `assert` enough? I have no opinion on this, but just want to make sure that we've thought about it!

Reply:
I'll leave the comment, but add a brief comment.

Reply:
OK.